### PR TITLE
stage_executor: Add support for inline `FROM --platform=` within Containerfile/Dockerfile

### PR DESCRIPTION
**Pull request by flouthoc**
See: https://docs.docker.com/engine/reference/builder/#from

Allows end users to configure executor's `OS`, `ARCH`,`VARIANT` via inline `--platform`.

Usage    
```
 FROM --platform=linux/arm64 alpine
 RUN uname -a
 ```
    
 ------
    
While this allows executor to pull base images with custom
`OS`, `ARCH`, `VARIANT` it still allows end-users to tag images with
 different format if they need to.
    
For example
    
Doing a `buildah build --platform linux/amd64 -t test .` on below containerfile.
    
 ```
 FROM --platform=linux/arm64 alpine
 RUN uname -a
 ```
    
Would result in tagging image as `linux/amd64` while `base` layer would
 be pulled for `linux/arm64`.
 
 Closes: https://github.com/containers/buildah/issues/3712
 Closes: https://github.com/containers/buildah/issues/3261

### CODE REVIEW

**Comment by flouthoc** [comment]
Please don't review this PR yet as this PR is a implementation of  : https://github.com/openshift/imagebuilder/pull/212 so lets wait for that.

**Comment by flouthoc** [comment]
@nalind @rhatdan PTAL. Also modified tests to also check if `qemu-user-static` is already there on platform as @edsantiago suggested in the last PR.

**Comment by flouthoc** [comment]
This PR closes two existing open issues.

**Comment by rhatdan** [comment]
LGTM
@nalind @vrothberg @TomSweeneyRedHat @ashley-cui @umohnani8 PTAL
 
As soon as this is merged, we can cut a new release.

**Review by rhatdan — APPROVED** [review]

File: tests/bud.bats
Comment (vrothberg) [inline tests/bud.bats:155]: Wouldn't that break as well without qemu-user-static?  I don't see the "trying" as suggested by the comment.

Comment (rhatdan) [reply tests/bud.bats:155]: Only need user-static if you are doing a RUN command. (Having not checked if this does a RUN.)

Comment (vrothberg) [reply tests/bud.bats:155]:
```
FROM --platform=linux/amd64 busybox
RUN uname -m
```

Comment (flouthoc) [reply tests/bud.bats:155]: ackd let me add same test here as well.

Comment (flouthoc) [reply tests/bud.bats:155]: I am removing `RUN` call so this does not fail.

File: tests/bud.bats
Comment (TomSweeneyRedHat) [inline tests/bud.bats:128]: I think otherarch will always be set to either amd64 or arm64 here.  Could we randomize this, or make the loop bigger and test all three of the other archs?

File: tests/bud.bats
Comment (vrothberg) [inline tests/bud.bats:170]: That would fail if no qemu isn't installed.  Don't we need the same checks as in the test abov?

Comment (flouthoc) [reply tests/bud.bats:170]: @vrothberg No this should not fail since we are only tagging as different `arch` the image which we are actually pulling is specifically set to match `host` arch. I am just creating `dockerfile` dynamically for specific host.

Comment (vrothberg) [reply tests/bud.bats:170]: The `RUN` instruction is still there, and we ignore the exit code.  If the build fails, there's no tagged image and then the inspect should fail.

File: tests/bud.bats
Comment (vrothberg) [inline tests/bud.bats:165]: This overwrite an existing Dockerfile in the git tree, please don't.

File: tests/bud.bats
Comment (flouthoc) [reply tests/bud.bats:170]: @vrothberg I'm confused a bit I guess. But the idea is if we pull the image which is equivalent to `hostarch` `RUN` should never fail.

Comment (vrothberg) [reply tests/bud.bats:170]: We build for a non-local platform, so `build` could fail if qemu isn't installed, couldn't it?  If it's intended to succeed, I am confused why we ignore the exist code, see `?`.

File: tests/bud.bats
Comment (nalind) [reply tests/bud.bats:165]: Yeah, do not merge until this is changed.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:614]: `arch` and `variant` need to be managed together, else on a `linux/arm/v7` system or the like you could conceivably find yourself attempting to pull an image for `linux/386/v7`.

File: tests/bud.bats
Comment (nalind) [inline tests/bud.bats:156]: Might be simpler to grep the output of `buildah version`'s `BuildPlatform`, which provides the entire value you're looking to use.

File: tests/bud.bats
Comment (flouthoc) [reply tests/bud.bats:167]: changed to `${TESTDIR}`

File: tests/bud.bats
Comment (flouthoc) [reply tests/bud.bats:158]: @nalind I need to toggle  only `arch` at certain places that's why i preferred this. Hope this is not a blocker.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: fixed.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:606]: Change this to `||`.  For non-ARM platforms, variant will typically be empty, and that would cause `arch` to be ignored for all of them.

File: tests/bud.bats
Comment (flouthoc) [reply tests/bud.bats:169]: @vrothberg I have removed the `?` so if build fails the test will break. I think extra `?` was added cause i copied this test from the first one.

Comment (nalind) [reply tests/bud.bats:169]: The `--platform` on the FROM instruction implicitly overrides the `--platform` we get at the command line, and this looks like it's attempting to verify that the build succeeds when the command line `--platform` points to a foreign architecture, but `FROM --platform` specifies the local one.  It'll succeed either way if emulation is enabled, though, which is probably not what we want.

Comment (nalind) [reply tests/bud.bats:169]: Ah, and that's why it checks on the next line.

Comment (flouthoc) [reply tests/bud.bats:169]: @nalind That's correct.

Comment (vrothberg) [reply tests/bud.bats:169]: > The --platform on the FROM instruction implicitly overrides the --platform we get at the command line, and this looks like it's attempting to verify that the build succeeds when the command line --platform points to a foreign architecture, but FROM --platform specifies the local one. It'll succeed either way if emulation is enabled, though, which is probably not what we want.

Is the behavior that the Dockerfile instructions overrides the CLI inherited from Docker or buildkit?  My expectation is that when I say `--platform foo` on the CLI that I get platform foo.

Comment (flouthoc) [reply tests/bud.bats:169]: @vrothberg Yes `buidlkit` supports that. The reason is so users could do `multi-platform` builds otherwise the variables which we set with cli `--platform` like `BUILDPLATFORM` and `TARGETPLATFORM` would not be a lot of useful to end users if they can't override base images.

There is more discussed here: https://github.com/containers/buildah/issues/3261 and https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images

Tryout example with `buildkit`:

```dockerfile
FROM --platform=linux/amd64 alpine
RUN uname -m
```

But build this docker file using `--platform` different than one set in `Dockerfile`.

```console
sudo docker buildx build --platform linux/arm64 -t test .
```

Now on inspecting this image `sudo docker inspect test`  you can actually see that image for overridden by `inline` `linux/amd64` platform but still tagged as `linux/arm64`.

**Pull request merged**
